### ISSUE TITLE
BatchNorm should use Bessel's correction consistently

### ISSUE THREAD

**Issue by Cyanogenoid** [pytorch/pytorch#1410]
Currently, Bessel's correction is applied to the running variance estimator in BatchNorm on [this line](https://github.com/pytorch/pytorch/blob/master/torch/lib/THCUNN/BatchNormalization.cu#L196) and the correction is not applied at train time. While the original paper for [BatchNorm](https://arxiv.org/abs/1502.03167) agrees with this, in the follow-up paper of the author, [Batch Renormalization](https://arxiv.org/abs/1702.03275), they state on page 2 when talking about BatchNorm during training:

> [...] the computation of the sample mean µ_B and sample standard deviation σ_B are part of the model architecture [...]

and on page 3 state:

> Batch Normalization, in fact, simply sets r = 1

where r is the ratio of minibatch standard deviation and estimated standard deviation, implying that during both training and testing, Bessel's correction is applied.

I think the approach for BatchNorm in Batch Renormalization makes more sense than unbiasing one variance but not the other, so I believe that PyTorch should allow an argument to BatchNorm that toggles this behaviour and perhaps consider switching the default. An argument would also bring visibility towards this implementation difference in Deep Learning frameworks.

Looking through some of the most popular DL libraries, I found no implementation that uses the unbiased variance during training and inference, but some that don't unbias either. Note that the former is essentially the same as the latter when the latter initializes gamma with the correction factor, so I imagine that they should behave mostly the same even without correcting gamma.
Some of the frameworks that don't unbias at all are: [Keras](https://github.com/fchollet/keras/blob/master/keras/layers/normalization.py#L182-L184), [Lasagne](https://github.com/Lasagne/Lasagne/blob/master/lasagne/layers/normalization.py#L297-L299), [TF-Slim](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/layers/python/layers/layers.py#L696-L697).
Some of course follow the original paper by unbiasing the running variance only: [Caffe](https://github.com/BVLC/caffe/blob/master/src/caffe/layers/batch_norm_layer.cpp#L143), [Theano](http://deeplearning.net/software/theano/library/tensor/nnet/bn.html).

Anecdotally, when biasing the running unbiased variances again by inverting Bessel's correction, I get statistically significant slightly better validation accuracies when using relatively small batch sizes (32) in densely connected networks.

**Comment by ezyang** [comment]
If someone wants to change this and it doesn't hurt performance, please do it, we'll accept the PR. (Also, someone needs to check if we haven't fixed this already.)

**Comment by Chillee** [comment]
I agree that it doesn't seem like it makes sense to use the correction at test time but not train time. After all, one of the points of batch norm is to make the neural network have an even distribution of inputs. If this distribution shifts between training and eval, that seems like a bad thing.

As for whether we should apply Bessel's consistently or not, that's less clear. On one hand, it seems like applying it consistently is the right thing to do, and could make a difference if you were constantly varying your batch size. On the other hand, it doesn't seem like any framework out there currently consistently applies Bessel's, and we might just want to fall in line.

PS: The status of this issue hasn't changed @ezyang. We still calculate training variance without it and test-time variance (running variance) with it: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/Normalization.cpp#L216

**Comment by ppwwyyxx** [comment]
> Looking through some of the most popular DL libraries, I found no implementation that uses the unbiased variance during training and inference

IIRC, TensorFlow's `tf.nn.fused_batch_norm` uses unbiased variance for EMA and biased variance for training (i.e. it agrees with PyTorch).

**Comment by wsshin** [comment]
I'm not PyTorch user, but landed on this thread from https://datascience.stackexchange.com/questions/24239/batch-normalization-variance-calculation.  

Could someone clarify what the OP's proposal is?  For reference, the original batch normalization paper used

- $μ_B = \frac{1}{m} ∑_{i=1}^m x_i$
- $σ_B^2 = \frac{1}{m} ∑_{i=1}^m (x_i - μ_B)^2$ 

during training for each mini-batch $B$, and

- $E[x] = E_B[μ_B]$
- $\mathrm{Var}[x] = \frac{m}{m-1} E_B[σ_B^2]$

during testing, where $E_B[⋅]$ is the expectation over all mini-batches.

Is the OP proposing to use 

- $μ_B = \frac{1}{m} ∑_{i=1}^m x_i$
- $σ_B^2 = \frac{1}{m-1} ∑_{i=1}^m (x_i - μ_B)^2$ 

during training for each mini-batch $B$, and

- $E[x] = E_B[μ_B]$
- $\mathrm{Var}[x] = \frac{N_B}{N_B-1} E_B[σ_B^2]$

where $N_B$ is the number of mini-batches?  I think that will be wrong.

**Comment by dhruvbird** [comment]
I'm not 100% sure about this but based on the docs. on the ONNX runtime https://onnx.ai/onnx/operators/onnx__BatchNormalization.html it seems like the population mean (division by `N`) is used by the ONNX BatchNorm operator even during inference, whereas it seems like PyTorch using the sample mean (division by `N-1`) during inference and the population mean during training. This could be a problem for my models exported to ONNX runtime. Please could someone verify this claim? What's the best way to work around this problem short of implementing one's own BatchNorm layer from scratch?

**Comment by 3f6a** [comment]
Some references:
- https://peterchng.com/blog/2023/08/21/batchnorm-and-the-curious-case-of-training-vs.-inference-variance/
- A. Karpathy lecture: https://www.youtube.com/watch?v=q8SA3rM6ckI&list=PLAqhIrjkxbuWI23v9cThsA9GvCAUhRvKZ&index=5&t=3917s
